src/cmap/auth/gssapi.ts

```diff
@@ -30,60 +51,161 @@ export class GSSAPI extends AuthProvider {
+              autoAuthorize: 1
+            },
+            (err, result) => {
+              if (err) return callback(err);
 
-        const command = {
-          saslStart: 1,
-          mechanism: 'GSSAPI',
-          payload,
-          autoAuthorize: 1
-        };
+              const doc = result.result;
+              negotiate(context!, doc.payload, (err, payload) => {
+                if (err) return callback(err);
 
-        connection.command('$external.$cmd', command, (err, result) => {
-          if (err) return callback(err, false);
+                externalCommand(
+                  {
+                    saslContinue: 1,
+                    conversationId: doc.conversationId,
+                    payload
+                  },
+                  (err, result) => {
+                    if (err) return callback(err);
 
-          const doc = result.result;
-          authProcess.transition(doc.payload, (err, payload) => {
-            if (err) return callback(err, false);
-            const command = {
-              saslContinue: 1,
-              conversationId: doc.conversationId,
-              payload
-            };
+                    const doc = result.result;
+                    finalize(context!, doc.payload, (err, payload) => {
+                      if (err) return callback(err);
 
-            connection.command('$external.$cmd', command, (err, result) => {
-              if (err) return callback(err, false);
+                      externalCommand(
+                        {
+                          saslContinue: 1,
+                          conversationId: doc.conversationId,
+                          payload
+                        },
+                        (err, result) => {
+                          if (err) return callback(err);
 
-              const doc = result.result;
-              authProcess.transition(doc.payload, (err, payload) => {
-                if (err) return callback(err, false);
-                const command = {
-                  saslContinue: 1,
-                  conversationId: doc.conversationId,
-                  payload
-                };
-
-                connection.command('$external.$cmd', command, (err, result) => {
-                  if (err) return callback(err, false);
-
-                  const response = result.result;
-                  authProcess.transition(null, err => {
-                    if (err) return callback(err);
-                    callback(undefined, response);
-                  });
-                });
+                          callback(undefined, result.result);
+                        }
+                      );
+                    });
+                  }
+                );
               });
-            });
-          });
+            }
+          );
         });
-      });
+      }
+    );
+  }
+}
+
+function initialize(
+  context: Omit<GSSAPIContext, 'client'>,
+  callback: (err?: Error, context?: GSSAPIContext) => void
```

Comment:
can't you use our `Callback` type here? Also, it looks like the signature is actually `Callback<void>` fwiw